api/handlers.go

```diff
@@ -546,21 +548,125 @@ func (api *API) userPUT(w http.ResponseWriter, req *http.Request, _ httprouter.P
+	if err != nil && !errors.Contains(err, database.ErrUserNotFound) {
+		api.WriteError(w, errors.New("failed to fetch user from the DB"), http.StatusInternalServerError)
+		return
+	}
+	if err == nil {
+		api.WriteError(w, errors.New("pubkey already registered"), http.StatusBadRequest)
+		return
+	}
+	ch, err := api.staticDB.NewChallenge(ctx, pk, database.ChallengeTypeUpdate)
+	if err != nil {
+		api.WriteError(w, err, http.StatusInternalServerError)
+		return
+	}
+	uu := &database.UnconfirmedUserUpdate{
+		Sub:         sub,
+		ChallengeID: ch.ID,
+		ExpiresAt:   ch.ExpiresAt.Truncate(time.Millisecond),
+	}
+	err = api.staticDB.StoreUnconfirmedUserUpdate(ctx, uu)
+	if err != nil {
+		api.WriteError(w, errors.AddContext(err, "failed to store unconfirmed user update"), http.StatusInternalServerError)
+		return
+	}
+	api.WriteJSON(w, ch)
+}
+
+// userAddPubKeyPOST updates the user's pubKey based on a challenge-response.
+func (api *API) userAddPubKeyPOST(w http.ResponseWriter, req *http.Request, _ httprouter.Params) {
+	sub, _, _, err := jwt.TokenFromContext(req.Context())
+	if err != nil {
+		api.WriteError(w, err, http.StatusUnauthorized)
+		return
+	}
+
+	ctx := req.Context()
+	// Get the challenge response.
+	var chr database.ChallengeResponse
+	err = chr.LoadFromRequest(req)
+	if err != nil {
+		api.WriteError(w, errors.AddContext(err, "missing or invalid challenge response"), http.StatusBadRequest)
+		return
+	}
+	pk, chID, err := api.staticDB.ValidateChallengeResponse(ctx, chr, database.ChallengeTypeUpdate)
+	if err != nil {
+		api.WriteError(w, errors.AddContext(err, "failed to validate challenge response"), http.StatusBadRequest)
+		return
+	}
+	// Check if the pubkey from the UnconfirmedUserUpdate is already associated
+	// with a user. That might have happened between the challenge creation and
+	// the current moment.
+	pku, err := api.staticDB.UserByPubKey(ctx, pk)
+	if err != nil && !errors.Contains(err, database.ErrUserNotFound) {
+		err = errors.AddContext(err, "failed to verify that the pubKey is not already in use")
+		api.WriteError(w, err, http.StatusInternalServerError)
+		return
+	}
+	if err == nil && pku.Sub != sub {
+		api.WriteError(w, errors.New("this pubKey already belongs to another user"), http.StatusBadRequest)
+		return
+	}
+	uu, err := api.staticDB.FetchUnconfirmedUserUpdate(ctx, chID)
+	if err != nil {
+		api.WriteError(w, errors.AddContext(err, "failed to fetch unconfirmed user update"), http.StatusInternalServerError)
+		return
+	}
+	if uu.Sub != sub {
+		api.staticLogger.Warnf("Potential attempt to modify another user's pubKey. Sub of challenge requester '%s', sub of response submitter '%s'", uu.Sub, sub)
+		api.WriteError(w, errors.New("user's sub doesn't match update sub"), http.StatusBadRequest)
+		return
+	}
+	u, err := api.staticDB.UserBySub(ctx, sub, false)
+	if err != nil {
+		api.WriteError(w, err, http.StatusInternalServerError)
+		return
+	}
+	u.PubKeys = append(u.PubKeys, pk)
```

Comment:
Do we want to allow multiple pubkeys per user?

Reply:
Yes, that was a specifically requested feature.